packages/plugin-kafka/src/kafka.ts

```diff
@@ -0,0 +1,149 @@
+import assert = require('assert')
+import { resolve } from 'path'
+import { NotFoundError } from 'common-errors'
+import { Microfleet, PluginTypes, LoggerPlugin } from '@microfleet/core'
+import { promisify } from 'util'
+import debugLog from 'debug'
+import { once } from 'events'
+
+import {
+  ProducerStream,
+  ConsumerStream,
+  createReadStream,
+  createWriteStream,
+} from 'node-rdkafka'
+
+import {
+  TopicConfig,
+  ProducerStreamOptions,
+  ConsumerStreamOptions,
+  KafkaConfig,
+} from './types'
+
+const debug = debugLog('plugin-kafka')
+
+/**
+ * Relative priority inside the same plugin group type
+ */
+export const priority = 0
+export const name = 'kafka'
+export const type = PluginTypes.transport
+
+/**
+ * Defines service extension
+ */
+export interface KafkaPlugin {
+  rdKafkaConfig: KafkaConfig
+  createConsumerStream: (
+    streamOptions: ConsumerStreamOptions,
+    conf?: KafkaConfig,
+    topicConf?: TopicConfig
+  ) => Promise<ConsumerStream>
+  createProducerStream: (
+    streamOptions: ProducerStreamOptions,
+    conf?: KafkaConfig,
+    topicConf?: TopicConfig
+  ) => Promise<ProducerStream>
+  close: () => Promise<void>
+}
+
+type AnyStream = ConsumerStream | ProducerStream
+
+export class KafkaFactory implements KafkaPlugin {
+  rdKafkaConfig: KafkaConfig
+  private _streams: Set<AnyStream>
+
+  constructor(config: KafkaConfig) {
+    this.rdKafkaConfig = config
+    this._streams = new Set<AnyStream>()
+  }
+
+  async connectStream(stream: AnyStream) {
+    const client = Object.prototype.hasOwnProperty.call(stream, 'consumer')
+    ? (stream as ConsumerStream).consumer
+    : (stream as ProducerStream).producer
+
+    const connectError = once(stream, 'error')
+    const connectSuccess = once(client, 'ready')
+
+    const [raceResult] = await Promise.race([connectError, connectSuccess])
+
+    if (raceResult instanceof Error) {
+      throw raceResult
+    }
```

Comment:
if changed to have separate client & then a stream - we dont need to listen to error/ready, we can simply await once(client 'ready') as it will throw when error is emitted

Reply:
Updated using promisified prototypes and splitted client & stream connection logic. 
But this breaks some rules, because node-rdkafka API docs says that 

> You should generally not make this class yourself, as it is not even exposed as part of module.exports. Instead, you should KafkaConsumer.createReadStream.